api/src/auth/drivers/saml.ts

```diff
@@ -0,0 +1,171 @@
+import * as validator from '@authenio/samlify-node-xmllint';
+import express, { Router } from 'express';
+import { parseJSON } from '@directus/shared/utils';
+import * as samlify from 'samlify';
+import logger from '../../logger';
+import { getAuthProvider } from '../../auth';
+import { COOKIE_OPTIONS } from '../../constants';
+import env from '../../env';
+import { RecordNotUniqueException } from '../../exceptions/database/record-not-unique';
+import {
+	InvalidCredentialsException,
+	InvalidProviderException,
+} from '../../exceptions';
+import { respond } from '../../middleware/respond';
+import { AuthenticationService, UsersService } from '../../services';
+import { AuthData, AuthDriverOptions, User } from '../../types';
+import asyncHandler from '../../utils/async-handler';
+import { getConfigFromEnv } from '../../utils/get-config-from-env';
+import { LocalAuthDriver } from './local';
+
+samlify.setSchemaValidator(validator);
+
+export class SAMLAuthDriver extends LocalAuthDriver {
+	idp: any;
+	sp: any;
+	usersService: UsersService;
+	config: Record<string, any>;
+
+	constructor(options: AuthDriverOptions, config: Record<string, any>) {
+		super(options, config);
+
+		this.config = config;
+		this.usersService = new UsersService({ knex: this.knex, schema: this.schema });
+		this.sp = samlify.ServiceProvider(getConfigFromEnv(`AUTH_${config.provider.toUpperCase()}_SP`));
+		this.idp = samlify.IdentityProvider(getConfigFromEnv(`AUTH_${config.provider.toUpperCase()}_IDP`));
+	}
+
+	async fetchUserID(identifier: string) {
+		const user = await this.knex
+			.select('id')
+			.from('directus_users')
+			.whereRaw('LOWER(??) = ?', ['external_identifier', identifier.toLowerCase()])
+			.first();
+
+		return user?.id;
+	}
+
+	async getUserID(payload: Record<string, any>) {
+		const { provider, emailKey, identifierKey, nameKey, givenNameKey, familyNameKey, allowPublicRegistration } =
+			this.config;
+
+		const email = payload[emailKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'];
+		const identifier = payload[identifierKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier'];
+		const userID = await this.fetchUserID(identifier);
+
+		const userInfo = {
```

Comment:
Is this object required? You could just put `given_name` and `family_name` into variables to be used with registration.

Reply:
This is also in keeping with our method auth methods more than anything, as we use the same code flow there as well.

Reply:
How do you mean? You create the `userInfo` object, but it's not actually used. The values `given_name` and `family_name` are just extracted during registration.

Reply:
We call the userInfo object a few lines down during the insert, that’s it really but it’s the same thing we do in other auth methods as well.

Reply:
In the other auth methods, the `userInfo` object is returned by the `claims` or `userinfo` calls, or by LDAP query, rather than being purposefully assembled. It's no biggie. It's just that it's somewhat redundant in this scenario.

Reply:
Yeah can look it over further on Monday 👍🏻